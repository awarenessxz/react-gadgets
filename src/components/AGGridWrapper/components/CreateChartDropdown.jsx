import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Checkbox from '@material-ui/core/Checkbox';
import Divider from '@material-ui/core/Divider';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Grid from '@material-ui/core/Grid';
import Input from '@material-ui/core/Input';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import Switch from '@material-ui/core/Switch';
import TextField from '@material-ui/core/TextField';
import Styles from '../AgGridWrapper.scss';

const CreateChartDropdown = props => {
    const [title, setTitle] = React.useState('');
    const [startRowIndex, setStartRowIndex] = React.useState(1);
    const [endRowIndex, setEndRowIndex] = React.useState(3);
    const [columnNames, setColumnNames] = React.useState(props.defaultColumnNames);
    const [isAllCellRange, setIsAllCellRange] = React.useState(true);
    const [showLegend, setShowLegend] = React.useState(true);
    const [legendPlacement, setLegendPlacement] = React.useState('right');
    const [chartType, setChartType] = React.useState('groupedColumn');
    const [chartPalette, setChartPalette] = React.useState('bright');
    const chartTypes = [
        'groupedColumn',
        'stackedColumn',
        'normalizedColumn',
        'groupedBar',
        'stackedBar',
        'normalizedBar',
        'line',
        'scatter',
        'bubble',
        'pie',
        'doughnut',
        'area',
        'stackedArea',
        'normalizedArea',
    ];
    const chartPalettes = ['borneo', 'material', 'pastel', 'bright', 'flat'];
    const legendPlacements = ['top', 'right', 'bottom', 'left'];

    const getColumnFieldsMapping = () => {
        return columnNames.map(value => {
            const index = props.defaultColumnNames.indexOf(value);
            return props.defaultColumnFields[index];
        });
    };

    const getParams = () => {
        return {
            cellRange: {
                rowStartIndex: isAllCellRange ? undefined : startRowIndex - 1,
                rowEndIndex: isAllCellRange ? undefined : endRowIndex - 1,
                columns: getColumnFieldsMapping(),
            },
            chartType: chartType,
            chartPalette: chartPalette,
            processChartOptions: function (params) {
                const opts = params.options;
                if (title.trim() !== '') {
                    opts.title.enabled = true;
                    opts.title.text = title;
                }
                opts.legend.enabled = showLegend;
                opts.legend.position = legendPlacement;
                opts.seriesDefaults.tooltip.renderer = function (params) {
                    const titleStyle = params.color
                        ? ' style="color: white; background-color:' + params.color + '"'
                        : '';
                    const title = params.title
                        ? '<div class="ag-chart-tooltip-title"' +
                          titleStyle +
                          '>' +
                          params.title +
                          '</div>'
                        : '';
                    const value = params.datum[params.yKey]
                        .toString()
                        .replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
                    return (
                        title +
                        '<div class="ag-chart-tooltip-content" style="text-align: center">' +
                        value +
                        '</div>'
                    );
                };
                return opts;
            },
        };
    };

    return (
        <div>
            <Grid container spacing={1}>
                <Grid item xs={12}>
                    <h4 style={{ margin: '5px 0' }}>Create Charts</h4>
                    <Divider />
                </Grid>
            </Grid>
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Chart Title
                </Grid>
                <Grid item xs={7}>
                    <TextField
                        id='chartTitle'
                        label='can be empty'
                        onChange={e => setTitle(e.target.value)}
                    />
                </Grid>
            </Grid>
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Columns
                </Grid>
                <Grid item xs={7}>
                    <FormControl className={Styles.inputControlLength}>
                        <Select
                            multiple
                            value={columnNames}
                            onChange={e => setColumnNames(e.target.value)}
                            input={<Input />}
                            renderValue={selected => selected.join(', ')}>
                            {props.defaultColumnNames.map((name, index) => {
                                return (
                                    <MenuItem key={index} value={name}>
                                        <Checkbox checked={columnNames.indexOf(name) > -1} />
                                        <ListItemText primary={name} />
                                    </MenuItem>
                                );
                            })}
                        </Select>
                    </FormControl>
                </Grid>
            </Grid>
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Cell Range
                </Grid>
                <Grid item xs={7}>
                    <FormControlLabel
                        control={
                            <Switch
                                checked={isAllCellRange}
                                onChange={() => setIsAllCellRange(!isAllCellRange)}
                                name='chartCellRange'
                            />
                        }
                        label={isAllCellRange ? 'All Rows' : 'Customized'}
                    />
                </Grid>
            </Grid>
            {!isAllCellRange && (
                <React.Fragment>
                    <Grid container spacing={3} alignItems='center'>
                        <Grid item xs={1} />
                        <Grid item xs={6}>
                            Start Row No.
                        </Grid>
                        <Grid item xs={5}>
                            <TextField
                                id='chartStartRow'
                                value={startRowIndex}
                                onChange={e => setStartRowIndex(parseInt(e.target.value))}
                                type='number'
                                style={{ maxWidth: '80px' }}
                            />
                        </Grid>
                    </Grid>
                    <Grid container spacing={3} alignItems='center'>
                        <Grid item xs={1} />
                        <Grid item xs={6}>
                            End Row No.
                        </Grid>
                        <Grid item xs={5}>
                            <TextField
                                id='chartEndRow'
                                value={endRowIndex}
                                onChange={e => setEndRowIndex(parseInt(e.target.value))}
                                type='number'
                                style={{ width: '80px' }}
                            />
                        </Grid>
                    </Grid>
                </React.Fragment>
            )}
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Chart Type
                </Grid>
                <Grid item xs={7}>
                    <FormControl className={Styles.inputControlLength}>
                        <Select value={chartType} onChange={e => setChartType(e.target.value)}>
                            {chartTypes.map((name, index) => {
                                return (
                                    <MenuItem key={index} value={name}>
                                        {name}
                                    </MenuItem>
                                );
                            })}
                        </Select>
                    </FormControl>
                </Grid>
            </Grid>
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Chart Palette
                </Grid>
                <Grid item xs={7}>
                    <FormControl className={Styles.inputControlLength}>
                        <Select
                            value={chartPalette}
                            onChange={e => setChartPalette(e.target.value)}>
                            {chartPalettes.map((name, index) => {
                                return (
                                    <MenuItem key={index} value={name}>
                                        {name}
                                    </MenuItem>
                                );
                            })}
                        </Select>
                    </FormControl>
                </Grid>
            </Grid>
            <Grid container spacing={3} alignItems='center'>
                <Grid item xs={5}>
                    Chart Legend
                </Grid>
                <Grid item xs={7}>
                    <FormControlLabel
                        control={
                            <Switch
                                checked={showLegend}
                                onChange={() => setShowLegend(!showLegend)}
                                name='chartLegend'
                            />
                        }
                        label={showLegend ? 'Show' : 'Hide'}
                    />
                </Grid>
            </Grid>
            {showLegend && (
                <Grid container spacing={3} alignItems='center'>
                    <Grid item xs={1} />
                    <Grid item xs={6}>
                        Placement
                    </Grid>
                    <Grid item xs={5}>
                        <FormControl style={{ width: '80px' }}>
                            <Select
                                value={legendPlacement}
                                onChange={e => setLegendPlacement(e.target.value)}>
                                {legendPlacements.map((name, index) => {
                                    return (
                                        <MenuItem key={index} value={name}>
                                            {name}
                                        </MenuItem>
                                    );
                                })}
                            </Select>
                        </FormControl>
                    </Grid>
                </Grid>
            )}
            <Grid container spacing={1} justify='flex-end' style={{ marginTop: '20px' }}>
                <Button
                    variant='contained'
                    color='primary'
                    onClick={() => props.onClick(getParams())}>
                    Create
                </Button>
            </Grid>
        </div>
    );
};

CreateChartDropdown.defaultProps = {
    defaultColumnNames: [],
};

CreateChartDropdown.propTypes = {
    /** list of all column header in grid */
    defaultColumnNames: PropTypes.arrayOf(PropTypes.string),
    /** list of all column header field in grid */
    defaultColumnFields: PropTypes.arrayOf(PropTypes.string),
    /** callback */
    onClick: PropTypes.func.isRequired,
};

export default CreateChartDropdown;
