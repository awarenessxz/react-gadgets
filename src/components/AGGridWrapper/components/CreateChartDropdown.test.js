import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import CreateChartDropdown from './CreateChartDropdown';

// function to render component before each test
const renderComponent = props => {
    const defaultProps = {
        onClick: jest.fn(),
    };
    const merged = { ...defaultProps, ...props };
    return render(<CreateChartDropdown {...merged} />); // returns methods (refer to api)
};

// 1. Testing if component renders properly
describe('Testing if component renders properly', () => {
    // Snapshot Testing
    it('Snapshot Testing', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });
});
