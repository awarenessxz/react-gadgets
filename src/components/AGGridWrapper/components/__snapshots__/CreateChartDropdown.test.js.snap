// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if component renders properly Snapshot Testing 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <h4
          style="margin: 5px 0px;"
        >
          Create Charts
        </h4>
        <hr
          class="MuiDivider-root"
        />
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Chart Title
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="chartTitle"
            id="chartTitle-label"
          >
            can be empty
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="chartTitle"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Columns
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <div
          class="MuiFormControl-root inputControlLength"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              <span>
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Cell Range
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-2 Mui-checked"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-4 MuiSwitch-input"
                  name="chartCellRange"
                  type="checkbox"
                  value=""
                />
                <span
                  class="MuiSwitch-thumb"
                />
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            All Rows
          </span>
        </label>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Chart Type
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <div
          class="MuiFormControl-root inputControlLength"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              groupedColumn
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="groupedColumn"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Chart Palette
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <div
          class="MuiFormControl-root inputControlLength"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              bright
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="bright"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        Chart Legend
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
      >
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-2 Mui-checked"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-4 MuiSwitch-input"
                  name="chartLegend"
                  type="checkbox"
                  value=""
                />
                <span
                  class="MuiSwitch-thumb"
                />
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Show
          </span>
        </label>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        Placement
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
      >
        <div
          class="MuiFormControl-root"
          style="width: 80px;"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              right
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="right"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-xs-flex-end"
      style="margin-top: 20px;"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Create
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
